### DISCUSSION TITLE
Converting Neper Mesh Format to Dream3D Inl Format

### DISCUSSION THREAD

**Discussion by PengWei97** [neperfepx/neper#698] [Q&A] [answered]
Dear Neper experts

I am writing to inquire about whether the mesh format output by `neper -M` can be exported to the `inl` format used in Dream3D/MOOSE, or if there is a method to convert tess mesh files to inl format.

I am using neper for material simulations and would like to import the generated mesh into Dream3D for further analysis and processing. However, I have found that the mesh format output by neper -M is not compatible with the inl format supported by Dream3D. Therefore, I am wondering if there is a way to convert the mesh generated by neper to inl format, or if there are other file formats that can be used to convert between neper and Dream3D.

I would greatly appreciate any help and advice that you can provide, and I hope to hear from you soon.

Thank you!

Best regards,

wei

**Comment by rquey** [comment]
I have no clue on this one, but maybe someone out there knows how to do it.

**Reply by PengWei97** [reply]
Thank you for your response. To clarify my request, I am trying to convert a mesh file in msh format to an inl file format. Specifically, I need to obtain the coordinates of each node, the crystal orientation (in terms of Euler angles), and the corresponding grain ID for each node.

Currently, I have only been able to locate the node coordinate information in the msh file, but I have not been able to find the Euler angles and grain ID information for each result. If a mapping between {phi1 PHI phi2 x y z FeatureId} can be established, I believe that I can write a script to achieve my goal.

![image](https://github.com/neperfepx/neper/assets/56183775/949516ae-b5ba-4b8a-be64-c2056f663f42)

**Reply by rquey** [reply]
Okay, so this is a simple format.  Maybe a good opportunity to add it to Neper. The relevant code is under [`src/neut/neut_mesh/neut_mesh_fprintf`](https://github.com/neperfepx/neper/tree/main/src/neut/neut_mesh/neut_mesh_fprintf).

**Reply by smallrain1996** [reply]
![image](https://github.com/neperfepx/neper/assets/52028117/b463defa-ef59-4f03-a930-a930d4c058fa)

**Reply by PengWei97** [reply, marked as answer]
Dear @rquey and @smallrain1996,

I have successfully completed the conversion from .msh to .inl files. I followed your suggested commands in neper to create the necessary files, including .msh, .stelset, and .stdnode. Additionally, I have written a script called "msh_to_inl.sh" to perform the conversion.

However, I haven't yet implemented the suggested code addition to neper due to my limited familiarity with its internal logic. Nevertheless, I'm happy to share my approach and the MATLAB script I developed. It would be greatly appreciated if you or someone else could incorporate my operations into neper for a seamless conversion from .msh to .inl files.

First, please find the commands I used to create the required files in neper:
```bash
neper -T -dim 2 -n "10::msfile(n_msfile)" -morpho "gg::msfile(morpho_msfile)" -domain "square(10,10)" -o d -ori "uniform" -oridescriptor 'euler-bunge:passive'
neper -V d.tess -datacellcol "scaleid(1)" -datacelltrs 0.1 -print img1

neper -M d.tess -statelset id,nodes -statnode id,x,y,z -elttype "quad" -cl 0.1
```
Second, I have attached the "[msh_to_inl.m.txt](https://github.com/neperfepx/neper/files/11538875/msh_to_inl.m.txt)" script for your reference.

Also, below is a diagram of the microarchitecture that I imported into moose to represent what I have accomplished.
![image](https://github.com/neperfepx/neper/assets/56183775/0f18c23a-dd8d-43f2-bb05-3a79d74705e4)


Please let me know if you have any further questions or if there is anything else I can assist you with.

Thank you for your understanding and support.

Best regards,
wei
